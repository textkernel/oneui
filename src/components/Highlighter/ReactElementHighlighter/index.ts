export {
    ReactElementHighlighter,
    Props as ReactElementHighlighterProps,
    ReactElementHighlighterTermResult,
} from './ReactElementHighlighter';
