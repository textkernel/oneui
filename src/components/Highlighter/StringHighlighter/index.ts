export {
    StringHighlighter,
    Props as StringHighlighterProps,
    HighlighterRenderer,
} from './StringHighlighter';
