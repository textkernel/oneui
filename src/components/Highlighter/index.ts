export * from './ReactElementHighlighter';
export * from './StringHighlighter';
