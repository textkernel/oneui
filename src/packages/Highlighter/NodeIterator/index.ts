export { NodeIterator, NodeIteratorOptions } from './NodeIterator';
