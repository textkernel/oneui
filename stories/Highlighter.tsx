import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select, text, withKnobs } from '@storybook/addon-knobs';
import {
    StringHighlighter,
    ReactElementHighlighter,
    HighlighterCoreAccuracy,
} from '@textkernel/oneui';

const HIGHLIGHT_STYLES = {
    backgroundColor: 'lightyellow',
} as CSSStyleDeclaration;

storiesOf('Atoms/Highlighter', module)
    .addDecorator(withKnobs)
    .add('StringHighlighter', () => {
        const ACCURACY = ['exact', 'partial'];
        const someText = text(
            'Placeholder',
            'We are looking for php, java and javascript developer'
        );
        const keywords = text('keywords', 'javascript developer|php|java')
            .split('|')
            .filter((i) => i !== '');
        return (
            <StringHighlighter
                string={someText}
                searchTerms={keywords}
                highlighterCoreOptions={{
                    accuracy: select('Accuracy', ACCURACY, ACCURACY[0]) as HighlighterCoreAccuracy,
                    ignoreCase: boolean('ignoreCase', true),
                    ignoreDiacritics: boolean('ignoreDiacritics', true),
                }}
                highlightRenderer={({ substring, ...substringProps }) => (
                    <b {...substringProps} style={{ backgroundColor: 'lightyellow' }}>
                        {substring}
                    </b>
                )}
            />
        );
    })
    .add('ReactElementHighlighter', () => {
        const ACCURACY = ['exact', 'partial'];
        const isCustomStylesUsed = boolean('Use custom styles', false);
        const keywords = text(
            'keywords',
            'lorem|officia|dolore|dolor|irure dolor|qui officia deserunt'
        )
            .split('|')
            .filter((i) => i !== '');
        return (
            <ReactElementHighlighter
                key={`${keywords}${isCustomStylesUsed}`}
                searchTerms={keywords}
                highlightStyles={isCustomStylesUsed ? HIGHLIGHT_STYLES : undefined}
                highlighterCoreOptions={{
                    accuracy: select('Accuracy', ACCURACY, ACCURACY[0]) as HighlighterCoreAccuracy,
                    ignoreCase: boolean('ignoreCase', true),
                    ignoreDiacritics: boolean('ignoreDiacritics', true),
                }}
                onComplete={(highlighter, result) => {
                    console.log('highlighter', highlighter);
                    console.log('result', result);
                }}
            >
                <>
                    <h1>Lorem ipsum dolor sit amet</h1>
                    <p>
                        Lorem ipsum dolor sit amet, eiusmod tempor incididunt ut labore et dolore
                        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
                        laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
                        reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
                        pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
                        officia deserunt mollit anim id est laborum.
                    </p>
                </>
            </ReactElementHighlighter>
        );
    });
