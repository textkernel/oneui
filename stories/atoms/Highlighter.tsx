import * as React from 'react';
import { StringHighlighter, ReactElementHighlighter } from '@textkernel/oneui';

const HIGHLIGHT_STYLES = {
    backgroundColor: 'lightyellow',
} as CSSStyleDeclaration;

export default {
    title: 'Atoms/Highlighter',
    component: StringHighlighter,
    subcomponents: { ReactElementHighlighter },
    argTypes: {
        accuracy: { options: ['exact', 'partial'], control: 'inline-radio' },
    },
};

export const _StringHighlighter = (args) => (
    <StringHighlighter
        {...args}
        searchTerms={args.keywords.split('|').filter((i) => i !== '')}
        highlighterCoreOptions={{
            accuracy: args.accuracy,
            ignoreCase: args.ignoreCase,
            ignoreDiacritics: args.ignoreDiacritics,
        }}
    />
);
_StringHighlighter.args = {
    keywords: 'javascript developer|php|java',
    string: 'We are looking for php, java and javascript developer',
    accuracy: 'partial',
    ignoreCase: true,
    ignoreDiacritics: true,
    highlightRenderer: ({ substring, ...substringProps }) => (
        <b {...substringProps} style={{ backgroundColor: 'lightyellow' }}>
            {substring}
        </b>
    ),
};

export const _ReactElementHighlighter = (args) => {
    const words = args.keywords.split('|').filter((i) => i !== '');

    return (
        <ReactElementHighlighter
            {...args}
            key={`${words}${args.isCustomStylesUsed}`}
            searchTerms={words}
            highlightStyles={args.isCustomStylesUsed ? HIGHLIGHT_STYLES : undefined}
            highlighterCoreOptions={{
                accuracy: args.accuracy,
                ignoreCase: args.ignoreCase,
                ignoreDiacritics: args.ignoreDiacritics,
            }}
        />
    );
};
_ReactElementHighlighter.args = {
    keywords: 'lorem|officia|dolore|dolor|irure dolor|qui officia deserunt',
    isCustomStylesUsed: false,
    accuracy: 'exact',
    ignoreCase: true,
    ignoreDiacritics: true,
    onComplete: (highlighter, result) => {
        console.log('highlighter', highlighter);
        console.log('result', result);
    },
    children: (
        <>
            <h1>Lorem ipsum dolor sit amet</h1>
            <p>
                Lorem ipsum dolor sit amet, eiusmod tempor incididunt ut labore et dolore magna
                aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
                aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
                velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
                non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
            </p>
        </>
    ),
};
