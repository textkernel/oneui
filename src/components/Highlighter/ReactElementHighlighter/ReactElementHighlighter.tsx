import * as React from 'react';
import { HighlighterCoreOptions, Highlighter } from '../../../packages/Highlighter';

type ReactElementHighlighterTermResult = {
    term: string;
    styles: CSSStyleDeclaration;
};
export interface Props {
    /** Target DOM node for highlight searching */
    children: React.ReactElement;
    /** Array of words or RegExp that is used for search */
    searchTerms: string[];
    /** Specify HighlighterCore options */
    highlighterCoreOptions?: HighlighterCoreOptions;
    /** Specify styles for highlighted terms */
    highlightStyles?: CSSStyleDeclaration;
    /** Fired after finishing highlighting */
    onComplete?: (
        instance: Highlighter,
        result: [ReactElementHighlighterTermResult[], ReactElementHighlighterTermResult[]],
        rootNode: HTMLElement
    ) => void;
}

export const ReactElementHighlighter: React.FC<Props> = (props) => {
    const { children, searchTerms, highlighterCoreOptions, highlightStyles, onComplete } = props;
    const elementRef = React.createRef<HTMLDivElement>();
    const key = JSON.stringify({ searchTerms, ...highlighterCoreOptions });

    React.useEffect(() => {
        if (elementRef.current) {
            const highlighter = new Highlighter({
                root: elementRef.current,
            });
            const [matchedTerms, notMatchedTerms] = highlighter.find(
                searchTerms,
                highlighterCoreOptions,
                highlightStyles
            );
            if (onComplete) {
                onComplete(highlighter, [matchedTerms, notMatchedTerms], elementRef.current);
            }
        }
    }, [elementRef, onComplete, searchTerms, highlighterCoreOptions, highlightStyles]);

    return (
        <div key={key} ref={elementRef}>
            {children}
        </div>
    );
};

ReactElementHighlighter.displayName = 'ReactElementHighlighter';

ReactElementHighlighter.defaultProps = {
    highlighterCoreOptions: {
        accuracy: 'exact',
        ignoreDiacritics: true,
        ignoreCase: true,
    },
};
