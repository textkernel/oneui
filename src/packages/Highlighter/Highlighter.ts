import { NodeIterator } from './NodeIterator';
import { HighlighterCore, HighlighterCoreOptions } from './HighlighterCore';
import { HighlighterNode } from './HighlighterNode';

export type HighlighterResultItem = {
    term: string;
    styles: CSSStyleDeclaration;
    amount: number;
};

enum SelectionStatus {
    selected = 'selected',
    notSelected = 'not_selected',
}

type FindInNodeFunction = (node: Node, onMatch: (node: HTMLElement) => void) => void;

export interface HighlighterScrollArgs {
    nodeTopPosition: number;
    heightViewport: number;
    scrollTop: number;
}

const SCROLLABLE_NODE_STYLES = {
    scrollBehavior: 'smooth',
} as CSSStyleDeclaration;

const HIGHLIGHTED_NODE_STYLES = {
    borderRadius: '1px',
    display: 'inline-block',
    textIndent: 'initial',
} as CSSStyleDeclaration;

const SELECTED_NODE_STYLES = {
    boxShadow: '0 0 0 2px #0097d1',
    transform: 'scale(1.2)',
    transition: 'transform .2s',
    position: 'relative',
    zIndex: '1',
} as CSSStyleDeclaration;

const RESET_SELECTED_NODE_STYLES = {
    boxShadow: '',
    transform: 'scale(1)',
    transition: 'none',
    position: 'initial',
    zIndex: '0',
} as CSSStyleDeclaration;

export interface HighlighterOptions {
    /** Defines scope where you want find and scroll highlight keywords */
    root: HTMLElement;
    /** Function responsible for scrolling */
    scrollFunction?: (values: HighlighterScrollArgs) => void;
    /** Defines DOM element that has scrolling */
    scrollableElement?: HTMLElement;
}

const KEYWORD_ATTRIBUTE = 'data-highlight-keyword';
const SELECTED_ATTRIBUTE = 'data-highlight-selected';

export class Highlighter {
    private root: HTMLElement;

    private scrollFunction?: (values: HighlighterScrollArgs) => void;

    private scrollableElement?: HTMLElement;

    private nodeIterator: NodeIterator;

    /**
     * Method used for comparing position two nodes by the appearance
     * Compare vertical and horizontal position
     */
    static comparePositionTwoNodes(node: HTMLElement, nextNode: HTMLElement) {
        const nodeBoundaries = node.getBoundingClientRect();
        const nextNodeBoundaries = nextNode.getBoundingClientRect();
        const verticalCompares = nodeBoundaries.top - nextNodeBoundaries.top;
        const horizontalCompares = nodeBoundaries.left - nextNodeBoundaries.left;

        if (verticalCompares === 0) {
            return horizontalCompares;
        }

        return verticalCompares;
    }

    /**
     * Method used for sort nodes by appearance
     * Order from top to bottom and left to right
     */
    static sortNodesByAppearance(nodes: HTMLElement[]) {
        return nodes.sort(Highlighter.comparePositionTwoNodes);
    }

    static createHighlightWrapper(term: string, styles: CSSStyleDeclaration) {
        const node = document.createElement('mark');
        HighlighterNode.applyStylesForNode(node, styles);
        node.innerText = term;
        return node;
    }

    static generateHighlightStyles(index: number) {
        return {
            ...HIGHLIGHTED_NODE_STYLES,
            backgroundColor: HighlighterNode.generateColor(index),
        } as CSSStyleDeclaration;
    }

    constructor({
        root,
        scrollFunction,
        scrollableElement = window.document.documentElement,
    }: HighlighterOptions) {
        this.root = root;
        this.scrollFunction = scrollFunction;
        this.scrollableElement = scrollableElement;
        this.nodeIterator = new NodeIterator({
            root,
            whatToShow: NodeFilter.SHOW_TEXT,
        });

        HighlighterNode.applyStylesForNode(scrollableElement, SCROLLABLE_NODE_STYLES);
    }

    private keywordNodes(keyword: string) {
        return this.root.querySelectorAll<HTMLElement>(`[${KEYWORD_ATTRIBUTE}="${keyword}"]`);
    }

    private get selectedNode() {
        return this.root.querySelector<HTMLElement>(
            `[${SELECTED_ATTRIBUTE}="${SelectionStatus.selected}"]`
        );
    }

    private triggerSelection(keyword: string) {
        let nextIndex = 0;
        const keywordNodes = Array.from(this.keywordNodes(keyword));
        const prevNode = this.selectedNode;

        if (prevNode) {
            HighlighterNode.applyStylesForNode(prevNode, RESET_SELECTED_NODE_STYLES);
        }

        const sortedKeywordNodes = Highlighter.sortNodesByAppearance(keywordNodes);

        if (prevNode) {
            const indexPrevNode = sortedKeywordNodes.indexOf(prevNode);
            prevNode.setAttribute(SELECTED_ATTRIBUTE, SelectionStatus.notSelected);

            nextIndex = indexPrevNode + 1;

            if (nextIndex >= sortedKeywordNodes.length) {
                nextIndex = 0;
            }
        }

        const nextNode = sortedKeywordNodes[nextIndex];

        if (nextNode) {
            HighlighterNode.applyStylesForNode(nextNode, SELECTED_NODE_STYLES);
            nextNode.setAttribute(SELECTED_ATTRIBUTE, SelectionStatus.selected);
        }
    }

    private scrollToSelectedKeyword() {
        const { selectedNode } = this;
        if (selectedNode && this.scrollableElement) {
            const nodeTopPosition = selectedNode.getBoundingClientRect().top;
            const heightViewport = this.scrollableElement.clientHeight || 0;
            const scrollTop = this.scrollableElement.scrollTop || 0;

            if (this.scrollFunction) {
                this.scrollFunction({
                    nodeTopPosition,
                    heightViewport,
                    scrollTop,
                });
            } else {
                this.scrollableElement.scrollTo(
                    0,
                    nodeTopPosition + scrollTop - heightViewport / 2
                );
            }
        }
    }

    private findKeywordInNodes(keyword: string, findInNode: FindInNodeFunction) {
        const iterator = this.nodeIterator.create();

        iterator.iterateThroughNodes((node) => {
            findInNode(node, (highlightedNode) => {
                highlightedNode.setAttribute(KEYWORD_ATTRIBUTE, keyword);
            });
        });
    }

    public find(
        terms: string[],
        highlighterCoreOptions?: HighlighterCoreOptions,
        highlightStyles?: CSSStyleDeclaration
    ) {
        const matchedTerms: HighlighterResultItem[] = [];
        const notMatchedTerms: HighlighterResultItem[] = [];
        const sortedTerms = HighlighterCore.sortTerms(terms);

        sortedTerms.forEach((term, index) => {
            let amount = 0;
            const styles = highlightStyles || Highlighter.generateHighlightStyles(index);
            const highlighterNode = new HighlighterNode({
                highlighterCore: new HighlighterCore([term], {
                    ...highlighterCoreOptions,
                }),
                highlightRenderer: (highlightTerm) => {
                    return Highlighter.createHighlightWrapper(highlightTerm, styles);
                },
            });
            this.findKeywordInNodes(term, (node, onMatch) => {
                highlighterNode.find(node, (highlightedNode) => {
                    amount += 1;
                    onMatch(highlightedNode);
                });
            });

            const termResultItem: HighlighterResultItem = {
                term,
                styles,
                amount,
            };

            if (amount > 0) {
                matchedTerms.push(termResultItem);
            } else {
                notMatchedTerms.push(termResultItem);
            }
        });

        const sortedMatchTerms = matchedTerms.sort((a, b) => b.amount - a.amount);

        return [sortedMatchTerms, notMatchedTerms];
    }

    public selectKeyword(keyword: string) {
        this.triggerSelection(keyword);
        this.scrollToSelectedKeyword();
    }
}
