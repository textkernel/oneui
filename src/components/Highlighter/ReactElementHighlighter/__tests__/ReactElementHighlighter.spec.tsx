import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { Highlighter } from '../../../../packages/Highlighter';
import { ReactElementHighlighter } from '../ReactElementHighlighter';

describe('ReactElementHighlighter', () => {
    const keywords = ['lorem', 'officia', 'dolore', 'dolor', 'irure dolor', 'qui officia deserunt'];
    const renderComponent = (props) =>
        render(
            <ReactElementHighlighter
                searchTerms={keywords}
                highlightRenderer={(keyword) => {
                    const node = document.createElement('mark');
                    node.innerText = keyword;
                    return node;
                }}
                onComplete={() => {}}
                {...props}
            >
                <>
                    <h1>Lorem ipsum dolor sit amet</h1>
                    <p>
                        Lorem ipsum dolor sit amet, eiusmod tempor incididunt ut labore et dolore
                        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
                        laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
                        reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
                        pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
                        officia deserunt mollit anim id est laborum.
                    </p>
                </>
            </ReactElementHighlighter>
        );

    it('should highlight and return result', (done) => {
        renderComponent({
            onComplete: (highlighter, result, root) => {
                expect(root.outerHTML).toMatchSnapshot();
                expect(result[0]).toHaveLength(5);
                expect(result[1]).toHaveLength(1);
                expect(highlighter).toBeInstanceOf(Highlighter);

                done();
            },
        });
    });

    it('should highlight keywords with custom styles', (done) => {
        renderComponent({
            highlightStyles: {
                color: 'lightyellow',
            },
            onComplete: (highlighter, result, root) => {
                expect(root.outerHTML).toMatchSnapshot();

                done();
            },
        });
    });
});
