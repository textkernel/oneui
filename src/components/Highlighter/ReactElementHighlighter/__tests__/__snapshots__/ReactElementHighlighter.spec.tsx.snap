// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactElementHighlighter should highlight and return result 1`] = `"<div><h1><mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(255, 241, 107);\\" data-highlight-keyword=\\"lorem\\"></mark> ipsum <mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(229, 218, 201);\\" data-highlight-keyword=\\"dolor\\"></mark> sit amet</h1><p><mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(255, 241, 107);\\" data-highlight-keyword=\\"lorem\\"></mark> ipsum <mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(229, 218, 201);\\" data-highlight-keyword=\\"dolor\\"></mark> sit amet, eiusmod tempor incididunt ut labore et <mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(255, 196, 166);\\" data-highlight-keyword=\\"dolore\\"></mark> magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute <mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(214, 244, 129);\\" data-highlight-keyword=\\"irure dolor\\"></mark> in reprehenderit in voluptate velit esse cillum <mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(255, 196, 166);\\" data-highlight-keyword=\\"dolore\\"></mark> eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa <mark style=\\"border-radius: 1px; display: inline-block; text-indent: initial; background-color: rgb(175, 245, 226);\\" data-highlight-keyword=\\"qui officia deserunt\\"></mark> mollit anim id est laborum.</p></div>"`;

exports[`ReactElementHighlighter should highlight keywords with custom styles 1`] = `"<div><h1><mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"lorem\\"></mark> ipsum <mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"dolor\\"></mark> sit amet</h1><p><mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"lorem\\"></mark> ipsum <mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"dolor\\"></mark> sit amet, eiusmod tempor incididunt ut labore et <mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"dolore\\"></mark> magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute <mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"irure dolor\\"></mark> in reprehenderit in voluptate velit esse cillum <mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"dolore\\"></mark> eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa <mark style=\\"color: lightyellow;\\" data-highlight-keyword=\\"qui officia deserunt\\"></mark> mollit anim id est laborum.</p></div>"`;
